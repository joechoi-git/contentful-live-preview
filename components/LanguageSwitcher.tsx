import React, { useState, useEffect } from "react";
import { useChangeLocale, useCurrentLocale } from "../locales/client";

interface LanguageSwitcherProps {
    className?: string;
}

type Language = "en" | "es";

const LanguageSwitcher: React.FC<LanguageSwitcherProps> = ({ className }) => {
    const [language, setLanguage] = useState<string>("en");

    let currentLocale = undefined;
    let changeLocale = undefined;
    try {
        // eslint-disable-next-line react-hooks/rules-of-hooks
        currentLocale = useCurrentLocale();
        // eslint-disable-next-line react-hooks/rules-of-hooks
        changeLocale = useChangeLocale();
    } catch (error) {
        // console.log((error as Error)?.message);
    }

    useEffect(() => {
        const savedLanguage = localStorage.getItem("language");
        if (savedLanguage) {
            setLanguage(savedLanguage);
            if (savedLanguage !== currentLocale && changeLocale) {
                changeLocale(savedLanguage as Language);
            }
        }
    }, [changeLocale, currentLocale]);

    const handleLanguageChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
        const newLanguage = event.target.value;
        setLanguage(newLanguage);
        localStorage.setItem("language", newLanguage);
        if (changeLocale) {
            changeLocale(newLanguage as Language);
        }
    };

    return (
        <>
            <p>Locale: {currentLocale}</p>
            <p>Language: {language}</p>
            <select
                value={language}
                onChange={handleLanguageChange}
                className={`rounded border-solid border-4 p-2 border-accent bg-primary text-primary-content ${className}`}
            >
                <option value="en">English</option>
                <option value="es">Spanish</option>
            </select>
        </>
    );
};

export default LanguageSwitcher;
