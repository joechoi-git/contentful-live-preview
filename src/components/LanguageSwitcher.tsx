import React, { useState, useEffect } from "react";

interface LanguageSwitcherProps {
    className?: string;
}

const LanguageSwitcher: React.FC<LanguageSwitcherProps> = ({ className }) => {
    const [language, setLanguage] = useState<string>(
        process.env.NEXT_PUBLIC_CONTENTFUL_LOCALE || "en-US"
    );

    useEffect(() => {
        const savedLanguage = localStorage.getItem("language");
        if (savedLanguage) {
            setLanguage(savedLanguage);
        }
    }, []);

    const handleLanguageChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
        const newLanguage = event.target.value;
        setLanguage(newLanguage);
        localStorage.setItem("language", newLanguage);
    };

    return (
        <select
            value={language}
            onChange={handleLanguageChange}
            className={`rounded border-solid border-4 p-2 border-accent bg-primary text-primary-content ${className}`}
        >
            <option value="en-US">English</option>
            <option value="es-US">Spanish</option>
        </select>
    );
};

export default LanguageSwitcher;
