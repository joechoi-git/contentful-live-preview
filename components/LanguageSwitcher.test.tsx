import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import LanguageSwitcher from "./LanguageSwitcher";

// Mock localStorage
const localStorageMock = (() => {
    let store: Record<string, string> = {};

    return {
        getItem(key: string) {
            return store[key] || null;
        },
        setItem(key: string, value: string) {
            store[key] = value;
        },
        clear() {
            store = {};
        },
        removeItem(key: string) {
            delete store[key];
        }
    };
})();

Object.defineProperty(window, "localStorage", {
    value: localStorageMock
});

describe("LanguageSwitcher", () => {
    beforeEach(() => {
        localStorage.clear();
    });

    test("loads language from local storage on mount", () => {
        localStorage.setItem("language", "es");
        render(<LanguageSwitcher />);
        expect(screen.getByDisplayValue("Spanish")).toBeInTheDocument();
    });

    test("changes language and stores it in local storage", () => {
        render(<LanguageSwitcher />);

        // Change to Spanish
        fireEvent.change(screen.getByRole("combobox"), { target: { value: "es" } });
        expect(screen.getByDisplayValue("Spanish")).toBeInTheDocument();
        expect(localStorage.getItem("language")).toBe("es");

        // Change back to English
        fireEvent.change(screen.getByRole("combobox"), { target: { value: "en" } });
        expect(screen.getByDisplayValue("English")).toBeInTheDocument();
        expect(localStorage.getItem("language")).toBe("en");
    });

    test("applies additional class names passed as props", () => {
        const { container } = render(<LanguageSwitcher className="extra-class" />);
        const selectElement = container.querySelector("select");
        expect(selectElement).toHaveClass("extra-class");
    });

    test("matches snapshot", () => {
        const { asFragment } = render(<LanguageSwitcher />);
        expect(asFragment()).toMatchSnapshot();
    });
});
