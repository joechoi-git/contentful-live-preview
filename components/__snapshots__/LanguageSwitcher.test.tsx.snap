// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSwitcher matches snapshot 1`] = `
<DocumentFragment>
  <p>
    Locale: 
  </p>
  <p>
    Language: en
  </p>
  <select
    class="rounded border-solid border-4 p-2 border-accent bg-primary text-primary-content undefined"
  >
    <option
      value="en"
    >
      English
    </option>
    <option
      value="es"
    >
      Spanish
    </option>
  </select>
</DocumentFragment>
`;
